File: python/insecure_patterns.py
"""Insecure crypto patterns — CipherRadar advantage (weakness detection)."""
import random
import string
import ssl
import hashlib
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
from cryptography.hazmat.primitives import hashes

# EXPECTED: Random | - | - | high | quantum-safe
# Weak PRNG for security-sensitive value
def weak_token():
    return random.randint(0, 999999)  # BAD: not cryptographically secure

# EXPECTED: Random | - | - | high | quantum-safe
# Weak PRNG for password generation
def weak_password(length=16):
    chars = string.ascii_letters + string.digits
    return ''.join(random.choice(chars) for _ in range(length))  # BAD

# EXPECTED: AES | CBC | 256 | critical | quantum-safe
# Hardcoded encryption key
HARDCODED_KEY = b"0123456789abcdef0123456789abcdef"  # BAD: hardcoded
def encrypt_hardcoded(plaintext):
    iv = b"\x00" * 16  # BAD: static IV
    cipher = Cipher(algorithms.AES(HARDCODED_KEY), modes.CBC(iv))
    enc = cipher.encryptor()
    return enc.update(plaintext) + enc.finalize()

# EXPECTED: AES | CBC | 256 | high | quantum-safe
# Static IV
STATIC_IV = b"\x00\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f"  # BAD
def encrypt_static_iv(key, plaintext):
    cipher = Cipher(algorithms.AES(key), modes.CBC(STATIC_IV))
    enc = cipher.encryptor()
    return enc.update(plaintext) + enc.finalize()

# EXPECTED: TLS | - | - | critical | quantum-vulnerable
# Disabled certificate verification
def insecure_ssl_context():
    ctx = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
    ctx.check_hostname = False  # BAD
    ctx.verify_mode = ssl.CERT_NONE  # BAD
    return ctx

# EXPECTED: PBKDF2 | - | - | critical | quantum-safe
# Iteration count in variable — very low
iterations = 10
def weak_kdf(password, salt):
    kdf = PBKDF2HMAC(algorithm=hashes.SHA256(), length=32, salt=salt, iterations=iterations)  # BAD
    return kdf.derive(password)

# EXPECTED: - | - | - | critical | quantum-safe
# Password in source code
DATABASE_PASSWORD = "SuperSecret123!"  # BAD: hardcoded password
API_KEY = "sk-proj-abc123def456ghi789"  # BAD: hardcoded API key

# EXPECTED: MD5 | - | - | high | broken
# MD5 for password hashing
def hash_password_md5(password):
    return hashlib.md5(password.encode()).hexdigest()  # BAD: never use MD5 for passwords

File: java/InsecurePatterns.java
package benchmark;

import javax.crypto.Cipher;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import javax.net.ssl.*;
import java.security.SecureRandom;
import java.security.cert.X509Certificate;
import java.util.Random;

public class InsecurePatterns {

    // EXPECTED: Random | - | - | high | quantum-safe
    // Weak PRNG used for security-sensitive value
    public byte[] weakPRNG() {
        Random rand = new Random(); // BAD: not cryptographically secure
        byte[] token = new byte[32];
        rand.nextBytes(token);
        return token;
    }

    // EXPECTED: SecureRandom | - | - | medium | quantum-safe
    // SecureRandom with hardcoded seed (reduces entropy)
    public SecureRandom seededSecureRandom() {
        SecureRandom sr = new SecureRandom(new byte[]{1, 2, 3, 4, 5}); // BAD: fixed seed
        return sr;
    }

    // EXPECTED: X509TrustManager | - | - | critical | quantum-vulnerable
    // Trust-all manager — disables certificate validation
    public void trustAllSetup() throws Exception {
        SSLContext ctx = SSLContext.getInstance("TLS");
        ctx.init(null, new TrustManager[]{
            new X509TrustManager() {
                public X509Certificate[] getAcceptedIssuers() { return new X509Certificate[0]; }
                public void checkClientTrusted(X509Certificate[] c, String a) {}
                public void checkServerTrusted(X509Certificate[] c, String a) {} // BAD
            }
        }, null);
    }

    // EXPECTED: HostnameVerifier | - | - | critical | quantum-vulnerable
    // Hostname verification bypass
    public void hostnameBypass() {
        HttpsURLConnection.setDefaultHostnameVerifier((hostname, session) -> true); // BAD
    }

    // EXPECTED: AES | ECB | - | high | quantum-safe
    // Static IV reuse
    private static final byte[] STATIC_IV = new byte[]{
        0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
        0x08, 0x09, 0x0A, 0x0B, 0x0C, 0x0D, 0x0E, 0x0F
    }; // BAD: static IV

    // EXPECTED: AES | CBC | - | high | quantum-safe
    public byte[] staticIVEncrypt(byte[] key, byte[] data) throws Exception {
        Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
        cipher.init(Cipher.ENCRYPT_MODE, new SecretKeySpec(key, "AES"), new IvParameterSpec(STATIC_IV)); // BAD
        return cipher.doFinal(data);
    }

    // EXPECTED: AES | ECB | - | high | quantum-safe
    // ECB mode for multi-block encryption
    public byte[] ecbMultiBlock(byte[] key) throws Exception {
        Cipher cipher = Cipher.getInstance("AES/ECB/PKCS5Padding"); // BAD
        cipher.init(Cipher.ENCRYPT_MODE, new SecretKeySpec(key, "AES"));
        byte[] largeData = new byte[1024]; // Multiple AES blocks
        return cipher.doFinal(largeData);
    }

    // EXPECTED: RSA | PKCS1 | - | medium | quantum-vulnerable
    // RSA without specifying padding (defaults to PKCS1)
    public byte[] rsaDefaultPadding(java.security.PublicKey pub, byte[] data) throws Exception {
        Cipher cipher = Cipher.getInstance("RSA"); // BAD: defaults to PKCS1Padding
        cipher.init(Cipher.ENCRYPT_MODE, pub);
        return cipher.doFinal(data);
    }
}
